### PR TITLE
simplify report generation

### PR DESCRIPTION
**Pull request by boris-dimitrov**

### CODE REVIEW

File: app/lib/report_helper.rb [hunk 1]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
...
-    cat_id_set = Set.new(lineage_info.map { |_, taxons| taxons[0]['superkingdom_taxid'] }).delete(nil)
-    if !cat_id_set.empty?
-      category_taxids = cat_id_set.sort.join(",")
-      taxon_name_arr = TaxonLineage.connection.select_all("select taxid, name from taxon_names where taxid in (#{category_taxids})").to_hash
-      cat_name_info = taxon_name_arr.group_by { |h| h['taxid'] }
-    else
-      cat_name_info = {}
+    tax2d.each do |_tax_id, tax_pair|
+      tax_pair['NT'] ||= zero_twin(tax_pair['NR'])
+      tax_pair['NR'] ||= zero_twin(tax_pair['NT'])
```
Comment (boris-dimitrov) [inline app/lib/report_helper.rb]: @cdebourcy

File: app/lib/report_helper.rb [hunk 2]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
...
+  def flip_type(t)
+    t == 'NR' ? 'NT' : 'NR'
+  end
+
+  def zero_twin(taxon)
+    result = taxon.clone
+    result['count'] = 0
+    result['rpm'] = 0
+    result['count_type'] = flip_type(result['count_type'])
+    result['zscore'] = -100
```
Comment (boris-dimitrov) [inline app/lib/report_helper.rb]: @cdebourcy per discussion with @yunfangjuan

File: app/lib/report_helper.rb [hunk 3]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
...
-      if level[tax_id] == species_level
-        species_nts = taxon_nts
-        species_nrs = taxon_nrs
-        genus_taxid = linfo["genus_taxid"]
-        genus_nts = tax_hit[[genus_taxid, 'NT']] || [nil]
-        genus_nrs = tax_hit[[genus_taxid, 'NR']] || [nil]
+      if view_level_int == TaxonCount::TAX_LEVEL_SPECIES
+        genus_taxid = tax_pair['NT']['genus_taxid']
+        # the genus_taxid should always be present in tax2d
+        # but sometimes it isn't... possibly due to a bug
```
Comment (boris-dimitrov) [inline app/lib/report_helper.rb]: confirmed by @yunfangjuan and workaround in pipeline DB loader pending

File: app/lib/report_helper.rb [hunk 4]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
...
-    summary = TaxonSummary.connection.select_all("select * from taxon_summaries where background_id = #{report.background.id}").to_hash
-    total_reads = report.pipeline_output.total_reads
+  def select_taxons(report)
     pipeline_output_id = report.pipeline_output.id
-    data = TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, count_type from taxon_counts where pipeline_output_id = #{pipeline_output_id}").to_hash
+    total_reads = report.pipeline_output.total_reads
+    background_id = report.background.id
+    # Note: stdev is never 0
+    TaxonCount.select("
+      taxon_counts.tax_id         AS  tax_id,
```
Comment (yunfangjuan) [inline app/lib/report_helper.rb]: beautiful~

File: app/lib/report_helper.rb [hunk 5]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
...
+      taxon_lineages.genus_taxid  AS  genus_taxid,
+      taxon_names.name            AS  name,
+      tgn.name                    AS  genus_name,
+      tcn.name                    AS  category_name,
+      taxon_counts.count          AS  count,
+      (count / #{total_reads}.0
+        * 1000000.0)              AS  rpm,
+      IF(
+        stdev IS NOT NULL,
+        GREATEST(-99, LEAST(99, (((count / #{total_reads}.0 * 1000000.0) - mean) / stdev))),
```
Comment (yunfangjuan) [inline app/lib/report_helper.rb]: add TODO for zscore truncation

File: app/lib/report_helper.rb (hunk 2 above)
Comment (yunfangjuan) [inline app/lib/report_helper.rb]: can you document the use of -100 here. even better define constant somewhere.

File: app/lib/report_helper.rb [hunk 6]
```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
     # to be extended for all taxonomic ranks when needed
   end
 
-  def compute_taxon_zscores(report)
-    summary = TaxonSummary.connection.select_all("select * from taxon_summaries where background_id = #{report.background.id}").to_hash
-    total_reads = report.pipeline_output.total_reads
+  def select_taxons(report)
```
Comment (yunfangjuan) [inline app/lib/report_helper.rb]: you don't just select taxons though?

File: app/lib/report_helper.rb (hunk 5 above)
Comment (boris-dimitrov) [reply app/lib/report_helper.rb]: @eggerdesigns what is the doc link where you track zscore truncation logic?  I'd like to add a comment cross-referencing that from this code location

File: app/lib/report_helper.rb (hunk 2 above)
Comment (boris-dimitrov) [reply app/lib/report_helper.rb]: good idea, will define z score special constants

**Pull request merged**